File: openmole/misc/org.openmole.misc.workspace/src/main/java/org/openmole/misc/workspace/internal/Workspace.java
package org.openmole.misc.workspace.internal;


import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.TreeMap;
import java.util.UUID;


import org.apache.commons.configuration.Configuration;
import org.apache.commons.configuration.ConfigurationException;
import org.apache.commons.configuration.FileConfiguration;
import org.apache.commons.configuration.PropertiesConfiguration;
import org.apache.commons.configuration.reloading.FileChangedReloadingStrategy;
import org.jasypt.exceptions.EncryptionOperationNotPossibleException;
import org.jasypt.util.text.BasicTextEncryptor;
import org.joda.time.Period;
import org.joda.time.format.ISOPeriodFormat;
import org.joda.time.format.PeriodFormatter;
import org.openmole.commons.aspect.caching.Cachable;
import org.openmole.commons.aspect.caching.ChangeState;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.misc.workspace.ConfigurationElement;
import org.openmole.misc.workspace.ConfigurationLocation;
import org.openmole.misc.workspace.IPasswordProvider;
import org.openmole.misc.workspace.IWorkspace;

public class Workspace implements IWorkspace {

    static final String group = Workspace.class.getSimpleName();
    static final String fixedPrefix = "file";
    static final String fixedPostfix = ".wf";
    static final String fixedDir = "dir";
    static final ConfigurationLocation passwordTest = new ConfigurationLocation(group, "passwordTest", true);
    static final String passwordTestString = "test";
    File location;
    Map<ConfigurationLocation, ConfigurationElement> configurations = new TreeMap<ConfigurationLocation, ConfigurationElement>();
    transient IPasswordProvider passwordProvider;
    transient long currentTime = System.currentTimeMillis();


    Workspace(File location) {
        this.location = location;
        /*if (!exist()) {
        create();
        }*/
        addToConfigurations(UniqueID, new ConfigurationElement() {

            @Override
            public String getDefaultValue() {
                return UUID.randomUUID().toString();
            }
        });
        addToConfigurations(ObjectRepoLocation, new ConfigurationElement() {

            @Override
            public String getDefaultValue() {
                return new File(getLocation(), DefaultObjectRepoLocaltion).getAbsolutePath();
            }
        });
        addToConfigurations(TmpLocation, new ConfigurationElement() {

            @Override
            public String getDefaultValue() {
                return new File(getLocation(), DefaultTmpLocation).getAbsolutePath();
            }
        });

        addToConfigurations(passwordTest, passwordTestString);
    }

    @Override
    public synchronized File getLocation() {
        if (!location.exists()) {
            location.mkdirs();
        }
        return location;
    }

    @Override
    public synchronized void addToConfigurations(ConfigurationLocation location, ConfigurationElement element) {
        configurations.put(location, element);
    }

    @Override
    public synchronized void addToConfigurations(ConfigurationLocation location, String element) {
        configurations.put(location, new ConfigurationElement(element));
    }

    @ChangeState
    @Override
    public synchronized void setLocation(File location) {
        this.location = location;
    }

    @Cachable
    synchronized TempDir getTmpDir() throws IOException, InternalProcessingError {
        File locTmp = new File(getPreference(TmpLocation), getPreference(UniqueID));
        locTmp = new File(locTmp, Long.toString(currentTime));

        if (!locTmp.exists()) {
            if (!locTmp.mkdirs()) {
                throw new IOException("Cannot create tmp dir " + locTmp.getAbsolutePath());
            }
        }

        locTmp.deleteOnExit();
        return new TempDir(locTmp);
    }


    @Override
    public File newTmpDir(String prefix) throws IOException, InternalProcessingError {
        return getTmpDir().createNewTempDir(prefix);
    }

    @Override
    public File newTmpFile(String prefix, String suffix) throws IOException, InternalProcessingError {
        return getTmpDir().createNewTempFile(prefix, suffix);
    }

    @Override
    public File newFile(String prefix, String suffix) throws IOException, InternalProcessingError {
        return getTmpDir().createNewFile(prefix, suffix);
    }

    public File getTmpFile(String name) throws IOException, InternalProcessingError {
        File ret = new File(getTmpDir().getLocation(), name);
        ret.deleteOnExit();
        return ret;
    }

    @Cachable
    private File getConfigurationFile() throws InternalProcessingError {
        File configurationCacheTmp;
        configurationCacheTmp = new File(getLocation(), ConfigurationFile);

        try {
            configurationCacheTmp.createNewFile();
        } catch (IOException e) {
            throw new InternalProcessingError(e, "Error creating the configuration file " + configurationCacheTmp.getAbsolutePath());
        }

        return configurationCacheTmp;
    }

    @Cachable
    private synchronized FileConfiguration getConfiguration() throws InternalProcessingError {
        try {
            FileConfiguration configuration = new PropertiesConfiguration(getConfigurationFile());
            configuration.setReloadingStrategy(new FileChangedReloadingStrategy());
            return configuration;
        } catch (ConfigurationException e) {
            throw new InternalProcessingError(e);
        }
    }


    @Override
    public String getDefaultValue(ConfigurationLocation location) {
        ConfigurationElement cf = configurations.get(location);
        if (cf == null) {
            return "";
        }
        String ret = cf.getDefaultValue();
        if (ret == null) {
            return "";
        }
        return ret;
    }

    @Override
    public synchronized String getPreference(ConfigurationLocation location) throws InternalProcessingError {
        Configuration conf = getConfiguration().subset(location.getGroup());
        String ret = conf.getString(location.getName());

        if (ret == null && configurations.containsKey(location)) {
            ret = configurations.get(location).getDefaultValue();
            setPreference(location, ret);
            return ret;
        }

        if (!location.isCyphered()) {
            return ret;
        } else {
            return getTextEncryptor().decrypt(ret);
        }

    }

    @Override
    public synchronized void setPreference(ConfigurationLocation location, String value) throws InternalProcessingError {
        try {
            Configuration conf = getConfiguration().subset(location.getGroup());
            String prop = location.isCyphered() ? getTextEncryptor().encrypt(value) : value;
            conf.setProperty(location.getName(), prop);
            getConfiguration().save();
        } catch (ConfigurationException ex) {
            throw new InternalProcessingError(ex);
        }
    }

    @Cachable
    private synchronized BasicTextEncryptor getTextEncryptor() throws InternalProcessingError {
        // if (textEncryptor == null) {
        String password = getPasswordProvider().getPassword();
        BasicTextEncryptor tmpTextEncryptor = new BasicTextEncryptor();
        tmpTextEncryptor.setPassword(password);
        //textEncryptor = tmpTextEncryptor;
        //  }

        return tmpTextEncryptor;
    }

    /*private synchronized void reinitTextEncryptor() {
    textEncryptor = null;
    }*/
    @Override
    public void setPasswordProvider(IPasswordProvider passwordProvider) {
        this.passwordProvider = passwordProvider;
    }

    private IPasswordProvider getPasswordProvider() {
        if (passwordProvider != null) {
            return passwordProvider;
        }

        synchronized (this) {
            if (passwordProvider == null) {
                passwordProvider = new SystemInPasswordProvider();
            }
            return passwordProvider;
        }
    }

    @Override
    public synchronized void removePreference(
            ConfigurationLocation location)
            throws InternalProcessingError {
        Configuration conf = getConfiguration().subset(location.getGroup());
        conf.clearProperty(location.getName());

    }

    @Override
    public File getFile(String name) throws IOException {
        return new File(getLocation(), name);
    }

    @Override
    public int getPreferenceAsInt(ConfigurationLocation location) throws InternalProcessingError {
        return new Integer(getPreference(location));
    }

    @Override
    public double getPreferenceAsDouble(ConfigurationLocation location) throws InternalProcessingError {
        return new Double(getPreference(location));
    }

    @Override
    public File newTmpFile() throws IOException, InternalProcessingError {
        return newTmpFile(fixedPrefix, fixedPrefix);
    }

    @Override
    public File newFile() throws IOException, InternalProcessingError {
        return newFile(fixedPrefix, fixedPostfix);
    }

    @Override
    public File newTmpDir() throws IOException, InternalProcessingError {
        return newTmpDir(fixedDir);
    }

    @ChangeState
    @Override
    public synchronized void resetPreferences() throws InternalProcessingError {
        getConfigurationFile().delete();
    }

    @Override
    public long getPreferenceAsLong(ConfigurationLocation location) throws InternalProcessingError {
        return new Long(getPreference(location));
    }

    @Override
    public void providePassword(String password) throws InternalProcessingError, UserBadDataError {
        setPasswordProvider(new StaticPasswordProvider(password));
        //reinitTextEncryptor();
        try {
            getPreference(passwordTest);
        } catch (EncryptionOperationNotPossibleException e) {
            throw new UserBadDataError(e, "Wrong password");
        }
    }


    @Override
    public long getPreferenceAsDurationInMs(ConfigurationLocation location) throws InternalProcessingError {
        return ISOPeriodFormat.standard().parsePeriod(getPreference(location)).toStandardSeconds().getSeconds() * 1000L;
    }

    @Override
    public int getPreferenceAsDurationInS(ConfigurationLocation location) throws InternalProcessingError {
        PeriodFormatter formatter = ISOPeriodFormat.standard();
        Period period = formatter.parsePeriod(getPreference(location));
        return period.toStandardSeconds().getSeconds();
    }
}

File: openmole/misc/org.openmole.misc.workspace/src/main/scala/org/openmole/misc/workspace/internal/Workspace.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.misc.workspace.internal

import java.io.File
import java.util.UUID
import java.util.logging.Level
import java.util.logging.Logger
import org.apache.commons.configuration.FileConfiguration
import org.apache.commons.configuration.PropertiesConfiguration
import org.apache.commons.configuration.reloading.FileChangedReloadingStrategy
import org.jasypt.exceptions.EncryptionOperationNotPossibleException
import org.jasypt.util.text.BasicTextEncryptor
import org.joda.time.format.ISOPeriodFormat
import org.openmole.commons.exception.InternalProcessingError
import org.openmole.commons.exception.UserBadDataError
import org.openmole.misc.workspace.ConfigurationLocation
import org.openmole.misc.workspace.IWorkspace
import org.openmole.commons.aspect.caching.Cachable;
import org.openmole.commons.aspect.caching.ChangeState;
import scala.collection.mutable.HashMap
import org.openmole.commons.tools.io.FileUtil

object Workspace {
  val group = "Workspace"
  val fixedPrefix = "file"
  val fixedPostfix = ".wf"
  val fixedDir = "dir"
  val passwordTest = new ConfigurationLocation(group, "passwordTest", true)
  val passwordTestString = "test"
}

class Workspace extends IWorkspace {

  import Workspace._
  import IWorkspace._ 
  
  var _location: File = null
  @transient private var textEncryptor = new BasicTextEncryptor
  val configurations = new HashMap[ConfigurationLocation, () => String]
    
  //@transient private var _password = ""

  this += (UniqueID, UUID.randomUUID.toString)
  this += (passwordTest, passwordTestString)
  
  @ChangeState
  override def location_= (location: File) = {
    new File(locationInternal, running).delete
    _location = location
    val run = new File(_location, running)
    run.createNewFile
    run.deleteOnExit
  }
  
  private def locationInternal = {
    if(_location == null) defaultLocation
    else _location
  }
  
  override def location: File = {
    val ret = locationInternal
    if (!ret.exists) ret.mkdirs
    ret
  }

  override def += (location: ConfigurationLocation, defaultValue: () => String) = synchronized {
    configurations(location) = defaultValue
  }

  override def += (location: ConfigurationLocation,  defaultValue: String) = synchronized {
    configurations(location) = () => defaultValue
  }

  @transient override lazy val sessionUUID = UUID.randomUUID
  
  @Cachable
  private[internal] def tmpDir: TempDir = {
    val tmpLocation = new File(new File(location, DefaultTmpLocation), sessionUUID.toString)

    if (!tmpLocation.mkdirs) {
      throw new InternalProcessingError("Cannot create tmp dir " + tmpLocation.getAbsolutePath)
    }
    
    Runtime.getRuntime.addShutdownHook(new Thread {
        override def run = FileUtil.recursiveDelete(tmpLocation)
      })
    
    new TempDir(tmpLocation)
  }

  override def newDir(prefix: String): File =  tmpDir.createNewDir(prefix)
  override def newFile(prefix: String, suffix: String): File =  tmpDir.createNewFile(prefix, suffix)

  @Cachable
  private def configurationFile: File = {
    val file = new File(location, ConfigurationFile)
    file.createNewFile
    file
  }

  @Cachable
  private def configuration: FileConfiguration = synchronized {
    val configuration = new PropertiesConfiguration(configurationFile)
    configuration.setReloadingStrategy(new FileChangedReloadingStrategy)
    return configuration
  }

  override def defaultValue(location: ConfigurationLocation): String = {
    configurations.get(location) match {
      case None => ""
      case Some(cf) => 
        val ret = cf()
        if (ret == null) ""
        else ret
    }
  }

  override def preference(location: ConfigurationLocation): String = synchronized {
    val conf = configuration.subset(location.group)
    val confVal = conf.getString(location.name)

    if (confVal == null) {
      configurations.get(location) match {
        case None => return null 
        case Some(value) =>
          val default = value()
          setPreference(location, default)
          return default
      }
    } 

    if (!location.cyphered) {
      return confVal
    } else {
      return textEncryptor.decrypt(confVal)
    }

  }

  override def setPreference(location: ConfigurationLocation, value: String) = synchronized {
    val conf = configuration.subset(location.group)
    val prop = if(location.cyphered) textEncryptor.encrypt(value) else value
    conf.setProperty(location.name, prop)
    configuration.save
  }

  /*@Cachable
  private def textEncryptor: BasicTextEncryptor = synchronized {
    val tmpTextEncryptor = new BasicTextEncryptor
    tmpTextEncryptor.setPassword(_password)
    tmpTextEncryptor
  }*/

  override def removePreference(location: ConfigurationLocation) = synchronized {
    val conf = configuration.subset(location.group)
    conf.clearProperty(location.name)
  }

  override def file(name: String): File = new File(location, name)

  override def preferenceAsInt(location: ConfigurationLocation): Int = preference(location).toInt
  
  override def preferenceAsDouble(location: ConfigurationLocation): Double = preference(location).toDouble
  
  override def newFile: File = newFile(fixedPrefix, fixedPostfix)
  
  override def newDir: File = newDir(fixedDir)

  @ChangeState
  override def reset = configurationFile.delete

  override def preferenceAsLong(location: ConfigurationLocation): Long = preference(location).toLong
  
  override def password_=(password: String) = synchronized {
    textEncryptor = new BasicTextEncryptor
    textEncryptor.setPassword(password)
    passwordIsCorrect //set preference
  }
  
  override def passwordIsCorrect = {
    try {
      preference(passwordTest)
      true
    } catch {
      case e => 
        Logger.getLogger(classOf[Workspace].getName).log(Level.FINE, "Password incorrect", e)
        false
    }
  }
  
  override def passwordChoosen = {
    isPreferenceSet(passwordTest)
  }

  override def preferenceAsDurationInMs(location: ConfigurationLocation): Long = {
    ISOPeriodFormat.standard.parsePeriod(preference(location)).toStandardSeconds.getSeconds * 1000L
  }

  override def preferenceAsDurationInS(location: ConfigurationLocation): Int = {
    val formatter = ISOPeriodFormat.standard
    val period = formatter.parsePeriod(preference(location))
    return period.toStandardSeconds.getSeconds
  }

  override def isPreferenceSet(location: ConfigurationLocation): Boolean = synchronized {
    val conf = configuration.subset(location.group)
    return (conf.getString(location.name) != null);
  }
}
